Objective-C/CBLDocument.mm

```diff
@@ -198,80 +201,102 @@ - (NSUInteger)countByEnumeratingWithState: (NSFastEnumerationState *)state
                                   objects: (id __unsafe_unretained [])buffer
                                     count: (NSUInteger)len
 {
-    return [_dict countByEnumeratingWithState: state objects: buffer count: len];
+    return [self.dict countByEnumeratingWithState: state objects: buffer count: len];
 }
 
 
 #pragma mark - SUBSCRIPTING
 
 
 - (CBLFragment*) objectForKeyedSubscript: (NSString*)key {
-    return [_dict objectForKeyedSubscript: key];
+    return [self.dict objectForKeyedSubscript: key];
 }
 
 
 #pragma mark - INTERNAL
 
 
+- (CBLDatabase*) database {
+    CBL_LOCK(self.lock) {
+        return _database;
+    }
+}
+
+
 - (void) setDatabase: (CBLDatabase *)database {
-    _database = database;
-    _c4db = _database.c4db;
+    CBL_LOCK(self.lock) {
+        _database = database;
+        _c4db = _database.c4db;
+    }
 }
 
 
 - (BOOL) isEmpty {
-    return _dict.isEmpty;
+    return self.dict.isEmpty;
 }
 
 
 - (BOOL) save: (NSError**)outError {
-    return [self saveWithConflictResolver: self.effectiveConflictResolver
-                                 deletion: NO
-                                    error: outError];
+    CBL_LOCK(self.lock) {
```

Comment:
I think the above sample does't cause the deadlock because only one thread can access the database at a time. The second thread doesn't hold the document's lock yet when waiting to access the databse to save the document.

I agree that it's dangerous to hold a lick while calling an unknown code. This is a difficult one to fix as
1. The database lock still needs to be hold until the save process is done.
2. The document's lock could be released during the conflict resolution but this could meanthat the document needs to be copied to send to the conflict handler. And any changes happen to the document from another thread while fixing conflicts will be ignored.

Could we discuss this during offline tomorrow?